Cache auth token in memory

jakobpn opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

msftclas commented:
@jakobpn,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

Zlatkovsky commented on src/client/public/auth.ts:
```diff
@@ -76,12 +76,14 @@ function proceedWithAuthInit(authRequest: AuthRequestParamData) {
                 const accessToken = authResponseKeyValues['access_token'];
+                const expiresIn = authResponseKeyValues['expires_in'];
                 if (accessToken) {
+                    const message = 'AUTH:access_token=' + accessToken + '&AUTH:expires_in=' + expiresIn;
```
Do you want to make this into a template string?

    const message = `AUTH:access_token=${accessToken}&AUTH:expires_in${expiresIn};

Zlatkovsky replied:
As far as the message goes, btw:  I was being super lazy last time, so hence I just did a key-value singleton.  But really it should just be a message, something like:

    {
        type: "auth",
        message: {
            accessToken: accessToken,
            expiresIn: expiresIn
        }
    }

Then stringify it and send it as a string.  And on the receiving end, check that the object is a string, do a try/catch around the JSON.stringify, and then do a check for whether the resulting object has a "type" property whose value is "auth".

Zlatkovsky commented on src/runtime-helpers/runtime-helpers.ts:
```diff
@@ -16,23 +16,36 @@
+        const cachedAccessToken = _getCachedAccessToken(clientId, resource);
+        if (cachedAccessToken) {
+            return Promise.resolve(cachedAccessToken);
         } else {
```
You don't need an "else" if your code does a "return" in the if statement.  It would never have reached the "else" statement stuff anyway, if it already returned out earlier.
So you can remove the "else" and un-indent that code block.

Zlatkovsky commented on src/runtime-helpers/runtime-helpers.ts:
```diff
@@ -213,6 +227,37 @@ module ScriptLab {
     }
+
+    function _extractAndCacheAccessToken(message: string, clientId: string, resource: string): string {
+        const authResponseKeyValues = message.split('&');
```
Please change to do JSON-parsing instead, since you'll be sending a JSON-stingified object (see comment above).  You can see why it's useful, too:  already this code is starting to make weird assumptions about the order of parameters, etc.  I was being lazy last time while it was only one parameter, though I probably should have known better even then.  But now that it's more than one parameter, it's time to refactor.

Zlatkovsky commented on src/runtime-helpers/runtime-helpers.ts:
```diff
@@ -213,6 +227,37 @@ module ScriptLab {
+        const authResponseKeyValues = message.split('&');
+        const accessToken = authResponseKeyValues[0].substr('AUTH:access_token='.length);
+        const expiresIn = Number(authResponseKeyValues[1].substr('AUTH:expires_in='.length));
+        cachedAuthTokens[clientId + resource] = { token: accessToken, expiry: Date.now() + ((expiresIn - 60) * 1000) };
```
Please explain the magic number "60".  E.g.,: `60 /* one minute of safety margin */` or whatever.
BTW, if it is in fact just one minute, I would make it 5 instead, just in case.  That's what Office.js helpers do, too.

Zlatkovsky commented on src/runtime-helpers/runtime-helpers.ts:
```diff
@@ -213,6 +227,37 @@ module ScriptLab {
+
+    function _getCachedAccessToken(clientId: string, resource: string): string {
+        const cachedAuthToken = cachedAuthTokens[clientId + resource];
+        if (cachedAuthToken) {
```
I prefer functions that bail out fast.  So instead of phrasing the `if` as a positive and then having an `else` that is a one-liner `return null`, I would reverse it to be:

    if (!cachedAuthToken) {
        return null;
    }

And then de-nest the rest of the logic.

Zlatkovsky commented on src/runtime-helpers/runtime-helpers.ts:
```diff
@@ -213,6 +227,37 @@ module ScriptLab {
+    }
+
+    function _removeCachedAccessToken(clientId: string, resource: string) {
+        if (cachedAuthTokens[clientId + resource]) {
```
Nit: do you want to create a helper method for concatenating.  I realize it seems silly, but what if you later want to add something else, or change the contactenation to have an underscore in-between (I would...).  Then you'll end up chasing the same logic all throughout the code, instead of changing only one function.

jakobpn commented:
Super comments. Thanks Michael!

Jakob

From: Michael Zlatkovsky [mailto:notifications@github.com]
Sent: Saturday, September 16, 2017 1:21 PM
To: OfficeDev/script-lab <script-lab@noreply.github.com>
Cc: Jakob Nielsen (OFFICE) <Jakob.Nielsen@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [OfficeDev/script-lab] Cache auth token in memory (#619)


@Zlatkovsky approved this pull request.

Looks good on the whole! See some comments, though. Marking as "Approved" so you can merge in yourself when you're done.

Note that Script Lab is still in questionable state (or rather, heartbeat refresh seems to be having issues... other things work, seemingly). Will look at it more tonight/tomorrow.

________________________________

In src/client/public/auth.ts<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOfficeDev%2Fscript-lab%2Fpull%2F619%23discussion_r139294348&data=02%7C01%7CJakob.Nielsen%40microsoft.com%7C363ef718d1e54ab8c32008d4fd407fe9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636411900860339793&sdata=an4bIq7cP31gu5eVk4Lc0Q7fFzHd3b82Ls6ecdgTFFc%3D&reserved=0>:

>                  if (accessToken) {

+                    const message = 'AUTH:access_token=' + accessToken + '&AUTH:expires_in=' + expiresIn;

Do you want to make this into a template string?

const message = `AUTH:access_token=${accessToken}&AUTH:expires_in${expiresIn};

________________________________

In src/client/public/auth.ts<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOfficeDev%2Fscript-lab%2Fpull%2F619%23discussion_r139294401&data=02%7C01%7CJakob.Nielsen%40microsoft.com%7C363ef718d1e54ab8c32008d4fd407fe9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636411900860339793&sdata=%2FK0J4CF3tJqmZduIo2tGdYurjtQuCZ7QtiTkjMtB%2Fm0%3D&reserved=0>:

>                  if (accessToken) {

+                    const message = 'AUTH:access_token=' + accessToken + '&AUTH:expires_in=' + expiresIn;

As far as the message goes, btw: I was being super lazy last time, so hence I just did a key-value singleton. But really it should just be a message, something like:

{

    type: "auth",

    message: {

        accessToken: accessToken,

        expiresIn: expiresIn

    }

}

Then stringify it and send it as a string. And on the receiving end, check that the object is a string, do a try/catch around the JSON.stringify, and then do a check for whether the resulting object has a "type" property whose value is "auth".

________________________________

In src/runtime-helpers/runtime-helpers.ts<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOfficeDev%2Fscript-lab%2Fpull%2F619%23discussion_r139294438&data=02%7C01%7CJakob.Nielsen%40microsoft.com%7C363ef718d1e54ab8c32008d4fd407fe9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636411900860339793&sdata=2%2FVUpsPr4DQic9TdBnywCosyGQjQGxAxelsdpuhN%2B88%3D&reserved=0>:

>      /** [PREVIEW] Gets an access token on behalf of the user for a particular service

      * @param resource: The resource provider (default: 'graph' = Microsoft Graph; but could also be custom URI)

     */

     export function getAccessToken(clientId: string, resource: string = 'graph'): Promise<string> {

-        if (_isPlainWeb()) {

-            return _getAccessTokenViaWindowOpen(clientId, resource);

-        } else if (_isDialogApiSupported()) {

-            return _getAccessTokenViaDialogApi(clientId, resource);

+        const cachedAccessToken = _getCachedAccessToken(clientId, resource);

+        if (cachedAccessToken) {

+            return Promise.resolve(cachedAccessToken);

         } else {

You don't need an "else" if your code does a "return" in the if statement. It would never have reached the "else" statement stuff anyway, if it already returned out earlier.
So you can remove the "else" and un-indent that code block.

________________________________

In src/runtime-helpers/runtime-helpers.ts<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOfficeDev%2Fscript-lab%2Fpull%2F619%23discussion_r139294540&data=02%7C01%7CJakob.Nielsen%40microsoft.com%7C363ef718d1e54ab8c32008d4fd407fe9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636411900860339793&sdata=c6VYJVu7KggedcjPNFVmZRVJUnWN41mBHYJ7Xz3yGbw%3D&reserved=0>:

> @@ -213,6 +227,37 @@ module ScriptLab {

     function _isDialogApiSupported() {

         return (window as any).Office.context.requirements.isSetSupported('DialogApi');

     }

+

+    function _extractAndCacheAccessToken(message: string, clientId: string, resource: string): string {

+        const authResponseKeyValues = message.split('&');

Please change to do JSON-parsing instead, since you'll be sending a JSON-stingified object (see comment above). You can see why it's useful, too: already this code is starting to make weird assumptions about the order of parameters, etc. I was being lazy last time while it was only one parameter, though I probably should have known better even then. But now that it's more than one parameter, it's time to refactor.

________________________________

In src/runtime-helpers/runtime-helpers.ts<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOfficeDev%2Fscript-lab%2Fpull%2F619%23discussion_r139294572&data=02%7C01%7CJakob.Nielsen%40microsoft.com%7C363ef718d1e54ab8c32008d4fd407fe9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636411900860339793&sdata=A7U857qt4Ch0C4wF3qtlcgmUsvkiKbOLYRXKA8ZXmGo%3D&reserved=0>:

> @@ -213,6 +227,37 @@ module ScriptLab {

     function _isDialogApiSupported() {

         return (window as any).Office.context.requirements.isSetSupported('DialogApi');

     }

+

+    function _extractAndCacheAccessToken(message: string, clientId: string, resource: string): string {

+        const authResponseKeyValues = message.split('&');

+        const accessToken = authResponseKeyValues[0].substr('AUTH:access_token='.length);

+        const expiresIn = Number(authResponseKeyValues[1].substr('AUTH:expires_in='.length));

+        cachedAuthTokens[clientId + resource] = { token: accessToken, expiry: Date.now() + ((expiresIn - 60) * 1000) };

Please explain the magic number "60". E.g.,: 60 /* one minute of safety margin */ or whatever.
BTW, if it is in fact just one minute, I would make it 5 instead, just in case. That's what Office.js helpers do, too.

________________________________

In src/runtime-helpers/runtime-helpers.ts<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOfficeDev%2Fscript-lab%2Fpull%2F619%23discussion_r139294591&data=02%7C01%7CJakob.Nielsen%40microsoft.com%7C363ef718d1e54ab8c32008d4fd407fe9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636411900860339793&sdata=btFRTHUxbH0%2FfYqZi33fN8Bh4Zok0AGULCoOFnO%2FptY%3D&reserved=0>:

> @@ -213,6 +227,37 @@ module ScriptLab {

     function _isDialogApiSupported() {

         return (window as any).Office.context.requirements.isSetSupported('DialogApi');

     }

+

+    function _extractAndCacheAccessToken(message: string, clientId: string, resource: string): string {

+        const authResponseKeyValues = message.split('&');

+        const accessToken = authResponseKeyValues[0].substr('AUTH:access_token='.length);

+        const expiresIn = Number(authResponseKeyValues[1].substr('AUTH:expires_in='.length));

+        cachedAuthTokens[clientId + resource] = { token: accessToken, expiry: Date.now() + ((expiresIn - 60) * 1000) };

+        return accessToken;

+    }

+

+    function _getCachedAccessToken(clientId: string, resource: string): string {

+        const cachedAuthToken = cachedAuthTokens[clientId + resource];

+        if (cachedAuthToken) {

I prefer functions that bail out fast. So instead of phrasing the if as a positive and then having an else that is a one-liner return null, I would reverse it to be:

if (!cachedAuthToken) {

    return null;

}

And then de-nest the rest of the logic.

________________________________

In src/runtime-helpers/runtime-helpers.ts<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOfficeDev%2Fscript-lab%2Fpull%2F619%23discussion_r139294618&data=02%7C01%7CJakob.Nielsen%40microsoft.com%7C363ef718d1e54ab8c32008d4fd407fe9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636411900860339793&sdata=rXR6ZP6cjML1DLCJh8lllgoM1GrMD5RNorbvzT%2Fr8j0%3D&reserved=0>:

> +            if (cachedAuthToken.expiry > Date.now()) {

+                /* Token is valid. Return it */

+                return cachedAuthToken.token;

+            }

+            else {

+                /* Token in cache has expired. Remove it from the cache */

+                _removeCachedAccessToken(clientId, resource);

+                return null;

+            }

+        } else {

+            return null;

+        }

+    }

+

+    function _removeCachedAccessToken(clientId: string, resource: string) {

+        if (cachedAuthTokens[clientId + resource]) {

Nit: do you want to create a helper method for concatenating. I realize it seems silly, but what if you later want to add something else, or change the contactenation to have an underscore in-between (I would...). Then you'll end up chasing the same logic all throughout the code, instead of changing only one function.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOfficeDev%2Fscript-lab%2Fpull%2F619%23pullrequestreview-63211054&data=02%7C01%7CJakob.Nielsen%40microsoft.com%7C363ef718d1e54ab8c32008d4fd407fe9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636411900860339793&sdata=DnjWc4B80Hn5jeuQXGzaBRcNYdihUUG2qI1MXZXt59Y%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAMgIOz-rPU9cH0541z_KEIMzcm5C2d4uks5sjC3BgaJpZM4PZ2TU&data=02%7C01%7CJakob.Nielsen%40microsoft.com%7C363ef718d1e54ab8c32008d4fd407fe9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636411900860339793&sdata=tNtPGRI3Bv7D9obaDXPJEaa9I5yZiAK0d6YRvZx7Vgw%3D&reserved=0>.
